doc/developer-guides/coding_guidelines.rst

```diff
@@ -695,6 +724,294 @@ Compliant example::
        }
 
 
+FN-11: The return type of a function shall be consistent
+========================================================
+
+The return type of a function shall be the same between its declaration and
+definition.
+
+Compliant example::
+
+    /* In `showcase.h`: */
+    uint32_t func_showcase(uint32_t param);
+    
+    /* In `showcase.c`: */
+    #include "showcase.h"
+    
+    uint32_t func_showcase(uint32_t param)
+    {
+        return param;
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* In `showcase.h`: */
+       uint64_t func_showcase(uint64_t param);
+       
+       /* In `showcase.c`: */
+       #include "showcase.h"
+       
+       uint32_t func_showcase(uint32_t param)
+       {
+               return param;
+       }
+
+
+FN-12: Banned functions shall not be used
+=========================================
+
+The following cases shall be covered:
+
+a) The following functions shall not be used, calloc, malloc, realloc, and free.
+   Dynamic memory allocation shall be replaced with static memory allocation;
+b) The following functions shall be used with restrictions, va_arg, va_start,
+   and va_end. Variadic functions shall only be used for the functions similar to
+   printf.
+
+Compliant example::
+
+    uint32_t showcase_array[32];
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       uint32_t *showcase_ptr = (uint32_t *)malloc(32U * sizeof(uint32_t));
+
+
+FN-13: A function with declaration shall have the corresponding definition
+==========================================================================
+
+Compliant example::
+
+    /* In `showcase.h`: */
+    /* declaration */
+    uint32_t func_showcase(uint32_t param);
+    
+    /* In `showcase.c`: */
+    #include "showcase.h"
+    
+    /* definition */
+    uint32_t func_showcase(uint32_t param)
+    {
+        return param;
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* In `showcase.h`: */
+       /* declaration */
+       uint32_t func_showcase(uint32_t param);
+       
+       /* There is no definition of `func_showcase` anywhere in all source files */
+
+
+FN-14: All defined functions shall be used
```

Comment:
This requirement confuses me a bit.  This would prohibit writing a library of useful functions (such as a math library) and then being required to remove functions that weren't called.  What's the intent of this requirement?  Must all code in the application have a chance to be executed (100% coverage)?

Reply:
This rule is to eliminate the unreachable code and improve the code maintainability.

An exception shall be added here. I will update this rule in v2.

Reply:
@dbkinder 
This rule is to eliminate the unreachable code and improve the code maintainability.
An exception shall be added here. I will update this rule in v2.